patterns/observers/CinemaTicketBookingExample

```diff
@@ -0,0 +1,90 @@
+import java.util.ArrayList;
+import java.util.List;
+
+// Defines the Subject interface that allows observers to register, unregister, and be notified.
+interface TicketSubject {
+    void registerObserver(TicketObserver observer);
+    void removeObserver(TicketObserver observer);
+    void notifyObservers();
+}
+
+// Implements the TicketSubject interface to manage ticket availability and observers.
+class TicketBookingSystem implements TicketSubject {
+    private List<TicketObserver> observers; // List to hold all registered observers.
+    private boolean ticketAvailable; // Flag to track ticket availability status.
+
+    public TicketBookingSystem() {
+        observers = new ArrayList<>(); // Initialize the list of observers.
+        ticketAvailable = false; // Initially, tickets are not available.
+    }
+
+    // Sets the ticket availability and notifies all observers about the change.
+    public void setTicketAvailable(boolean available) {
+        this.ticketAvailable = available; // Update ticket availability status.
+        notifyObservers(); // Notify all registered observers about the status change.
+    }
+
+    // Adds an observer to the list of observers.
+    @Override
+    public void registerObserver(TicketObserver observer) {
+        observers.add(observer); // Add the observer to the list.
+    }
+
+    // Removes an observer from the list of observers.
+    @Override
+    public void removeObserver(TicketObserver observer) {
+        observers.remove(observer); // Remove the observer from the list.
+    }
+
+    // Notifies all registered observers about the ticket availability status.
+    @Override
+    public void notifyObservers() {
+        for (TicketObserver observer : observers) { // Loop through each observer in the list.
+            observer.update(ticketAvailable); // Notify the observer with the current ticket availability status.
+        }
+    }
+}
+
+// Defines the Observer interface for receiving updates from the Subject.
+interface TicketObserver {
+    void update(boolean ticketAvailable);
```

Comment:
This isn't exactly wrong, but in Observer it might be more typical that the state change that triggers a notification would be on something more substantial than a boolean flag. So it might be better to pass some more generic data here (just so it's clear to readers that they're not limited to using a boolean).